### DISCUSSION TITLE
Referencing "targets" using Markdown

### DISCUSSION THREAD

**Discussion by LordOfDragons** [microsoft/vscode-discussions#1058] [Extension Development QnA] [answered]
For my language server I've done also parsing doc-string ("/** ... */" type comments). I'm using a subset of Doxygen for my language. In particular this means I'd like to include support for "\ref", "\see" and other tags which reference code elements in a project. For example this could be a reference to a class, class function or similar. I can resolve these information (document URI and range) but I could not find a way how to output text in Markdown (MarkupKind.markdown) to create a hyperlinked text pointing at the URI-range. How can this be done if it is actually possible to do.

**Comment by theReynald** [comment, marked as answer]
Hi @LordOfDragons.

To create hyperlinks in Markdown for your Language Server Protocol (LSP) responses, you need to format the text appropriately. Since you're using Markdown (specifically `MarkupKind.markdown`), you can include hyperlinks using the following syntax:

```markdown
[link text](URI)
```

Here's how you can generate Markdown with hyperlinks in your LSP server:

1. **Construct the Markdown String**: When you resolve information such as document URI and range for `\ref` or `\see` tags, construct the corresponding Markdown string with hyperlinks.

2. **Use Range Information**: If you have range information, you can use it to generate a link that navigates to a specific position in the document. For example, you can append `#L<line number>` to the URI to navigate to a specific line.

3. **Send Markdown Response**: Once you've constructed the Markdown string with hyperlinks, send it as part of your LSP response using `MarkupContent` with `MarkupKind.markdown`.

Here's an example of how you can do this in TypeScript:

```typescript
import { MarkupContent } from 'vscode-languageserver';

// Constructing Markdown string with a hyperlink
const linkText = 'Go to MyClass';
const documentUri = 'file:///path/to/your/document';
const lineNumber = 10; // Line number where MyClass is defined
const markdownString = `[${linkText}](${documentUri}#L${lineNumber})`;

// Creating MarkupContent with Markdown kind
const markupContent: MarkupContent = {
    kind: 'markdown',
    value: markdownString
};

// Sending the response to the client
connection.sendNotification('documentationResult', markupContent);
```

In this example, `documentationResult` is a notification that contains the Markdown content. Adjust this according to your LSP protocol and client-server communication.

By constructing Markdown strings with hyperlinks using the appropriate URI and range information, you can provide clickable references to code elements in your LSP responses.

Does that help?

**Reply by LordOfDragons** [reply]
That seems to work. What about column position? Locations have a line and column. Jumping to the start of the symbol would be nice.
